<a href="https://colab.research.google.com/github/Nuelky/FE/blob/main/MScFE_620_DP_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Step 1: RQuestions 5, 6, and 7 from GWP1 Using the Black-Scholes Model for European Options.

The Black-Scholes model-This is a widely used analytical model for pricing European-style options; it provides closed-form solutions for the fair value of call and put options by assuming constant volatility and continuous time, among other assumptions (Hull, 2018).

Here, we"ll apply the Black-Scholes formula to price a European call and a European put option with parameters provided in our GWP1.

### Question 5: Lets do the-Pricing European Call and Put Options

**We know the Parameters from GWP1:**

 Parameters based on the GWP1 for European options
S0 = 100  # Initial stock price
K = 100   # Strike price
r = 0.05  # Risk-free interest rate
T = 0.25  # Time to maturity in years (3 months)
sigma = 0.2  # Volatility (20%)

- Stock Price ($S_0$): 100 (This is the initial stock price from GWP1).
-Our Strike Price ($K$) is 100 for the calland 182 for put.
- The Risk-Free Rate ($r$) is (0.05).
- Our Volatility ($\sigma$)is 20% (0.2).
- Time to Maturity ($T$)= 0.25 years Time to maturity in years (3 months)

So we will use these parameters in the Black-Scholes formulas for pricing the European call and put options.
The Black-Scholes model will calculates the prices using the following closed-form formulas:

For a European call option:-

$$
C = S_0 \Phi(d_1) - K e^{-rT} \Phi(d_2)
$$

For a European put option:-

$$
P = K e^{-rT} \Phi(-d_2) - S_0 \Phi(-d_1)
$$

where:

$$
d_1 = \frac{\ln\left(\frac{S_0}{K}\right) + \left(r + \frac{\sigma^2}{2}\right)T}{\sigma \sqrt{T}}
$$

$$
d_2 = d_1 - \sigma \sqrt{T}
$$

In these equations above,

we have $\Phi$ which represents the cumulative distribution function of a standard normal variable whlie on the other side $S_0$ represents the underlying asset's price (McDonald, 2013).

Let use this as a code below


In [ ]:
import numpy as np
from scipy.stats import norm

# Black-Scholes pricing function
def black_scholes_price(S0, K, r, T, sigma, option_type="call"):
    """
    Calculate the Black-Scholes price for a European call or put option.

    Parameters:
    - S0: Initial stock price
    - K: Strike price
    - r: Risk-free interest rate
    - T: Time to maturity
    - sigma: Volatility
    - option_type: "call" or "put"

    Returns:
    - price: The Black-Scholes price of the option
    """
    d1 = (np.log(S0 / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
    d2 = d1 - sigma * np.sqrt(T)
    if option_type == "call":
        price = S0 * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2)
    elif option_type == "put":
        price = K * np.exp(-r * T) * norm.cdf(-d2) - S0 * norm.cdf(-d1)
    return price

# Updated parameters based on GWP1
S0 = 100     # Initial stock price
K = 100      # Strike price (same for both call and put in this case)
r = 0.05     # Risk-free interest rate
T = 0.25     # Time to maturity (3 months in years)
sigma = 0.2  # Volatility

# Calculate option prices using the updated parameters
call_price = black_scholes_price(S0, K, r, T, sigma, option_type="call")
put_price = black_scholes_price(S0, K, r, T, sigma, option_type="put")

print(f"European Call Option Price: {call_price:.2f}")
print(f"European Put Option Price: {put_price:.2f}")


European Call Option Price: 4.61
European Put Option Price: 3.37


**European Call Option Price (4.61)**

We have the calculated price of 4.61 for the European call option, which reflects its intrinsic and time value. Given the current stock price ($S_0 = 100$) equal to the strike price ($K = 100$), the call option is considered at the money. Although it does not possess intrinsic value at this moment, it holds time value and potential upside as the expiration date approaches. The Black-Scholes model, as applied here, accounts for factors such as the stock price relative to the strike price, a short time to expiration ($T = 0.25$ years), a moderate volatility ($\sigma = 20\%$), and a risk-free rate ($r = 5\%$) (Hull, 2018).

**European Put Option Price (3.37)**

The put option is priced at 3.37, reflecting its value under current market conditions. Since the stock price ($S_0 = 100$) equals the strike price ($K = 100$), this put option is also at the money, with no intrinsic value but some time value, providing potential downside protection until expiration. The Black-Scholes model suggests this relatively low premium due to the short time to maturity and moderate volatility, indicating the put option is primarily valued for its potential as a hedge against downside risk (McDonald, 2013).


### Question 6: Calculating Delta for European Call and Put Options

In our previous assignment, we know that Delta ($\Delta$) represents the sensitivity of the option price to changes in the underlying asset's price.

Here, In the Black-Scholes model, delta is calculated as follows-:

For a European call option=

$$
\Delta_{\text{call}} = \Phi(d_1)
$$

For a European put option=

$$
\Delta_{\text{put}} = \Phi(d_1) - 1
$$

where $d_1$ is calculated as in the pricing formulas above (Hull, 2018). Delta will provides insight into how much the option price would change with a $1 change in the underlying asset price and thus making it useful for hedging and risk management (McDonald, 2013).


In [ ]:
# Lets define the Black-Scholes delta calculation function
def black_scholes_delta(S0, K, r, T, sigma, option_type="call"):
    """
    Calculate the Black-Scholes delta for a European call or put option.

    Parameters:
    - S0: Initial stock price
    - K: Strike price
    - r: Risk-free interest rate
    - T: Time to maturity
    - sigma: Volatility
    - option_type: "call" or "put"

    Returns:
    - delta: The delta of the option
    """
    d1 = (np.log(S0 / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
    if option_type == "call":
        delta = norm.cdf(d1)
    elif option_type == "put":
        delta = norm.cdf(d1) - 1
    return delta

# Calculate deltas
call_delta = black_scholes_delta(S0, K_call, r, T_call, sigma, option_type="call")
put_delta = black_scholes_delta(S0, K_put, r, T_put, sigma, option_type="put")

print(f"Delta for European Call Option: {call_delta:.4f}")
print(f"Delta for European Put Option: {put_delta:.4f}")


Delta for European Call Option: 0.5695
Delta for European Put Option: -0.4305


In [ ]:
import numpy as np
from scipy.stats import norm

# Parameters for the GWP1 example
S0 = 100      # Initial stock price
K = 100       # Strike price
r = 0.05      # Risk-free interest rate
T = 0.25      # Time to maturity (3 months in years)
sigma = 0.2   # Volatility

# Calculate d1
d1 = (np.log(S0 / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))

# Delta for Call and Put Options
delta_call = norm.cdf(d1)
delta_put = norm.cdf(d1) - 1

# Output the results
print(f"Delta for European Call Option: {delta_call:.4f}")
print(f"Delta for European Put Option: {delta_put:.4f}")


Delta for European Call Option: 0.5695
Delta for European Put Option: -0.4305


### Comments on ---Delta Calculations for Sensitivity Analysis (Q6)

**The Delta of the European Call Option (0.5695)**-The Delta of 0.5695 will indicates that the European call option has moderate sensitivity to changes in the stock's price. A Delta of approximately 0.57 suggests that for each $1 increase in stock price, the call option’s price will increase by about 0.57. This moderate Delta=is characteristic of an at-the-money call option and it is  reflecting balanced potential gains and the probability of finishing in the money. Delta is a crucial metric in hedging because it indicates how much of the underlying asset should be bought or sold to maintain a neutral position (Hull, 2018).

Delta of the European Put Option (-0.4305)-The Delta of -0.4305 for the put option reflects moderate sensitivity to the underlying stock price with a negative sign which is indicating the inverse relationship typical of put options. A Delta of -0.43 will suggests that the put option price will decrease by about 0.43 for every $1 increase in the stock price. This value is (typical) for an at-the-money or slightly in-the-money put option, indicating that the option is moderately sensitive to price changes rather than being minimally sensitive (McDonald, 2013).


### Question 7: Calculating Vega for European Call and Put Options
What's Vega?

This is a measure of the sensitivity of the option's price to changes in the volatility of the underlying asset. Unlike the delta which will measures sensitivity to price, vega will assesse sensitivity to volatility which makes it essential for options with high volatility or longer time to maturity (Hull, 2018).

Vvega is calculated as follows in the Black-Scholes model-

$$
\text{Vega} = S_0 \sqrt{T} \, \phi(d_1)
$$

Where, $\phi(d_1)$ =is the probability density function of the standard normal distribution evaluated at $d_1$.


In [ ]:
# Lts do the Black-Scholes vega calculation function
def black_scholes_vega(S0, K, r, T, sigma):
    """
    This is to Calculate the Black-Scholes vega for a European option.

    Parameters:
    - S0: Initial stock price
    - K: Strike price
    - r: Risk-free interest rate
    - T: Time to maturity
    - sigma: Volatility

    Returns:
    - vega: The vega of the option
    """
    d1 = (np.log(S0 / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
    vega = S0 * norm.pdf(d1) * np.sqrt(T)
    return vega

# Lets Calculate vegas
call_vega = black_scholes_vega(S0, K_call, r, T_call, sigma)
put_vega = black_scholes_vega(S0, K_put, r, T_put, sigma)

print(f"Vega for European Call Option= {call_vega:.4f}")
print(f"Vega for European Put Option= {put_vega:.4f}")


Vega for European Call Option= 19.6440
Vega for European Put Option= 19.6440


### Comments on---Vega Calculations for Volatility Sensitivity (Q7)

The Vega of 19.64 for the European options-this indicates that the option price is moderately sensitive to changes in volatility. This will means that for every 1% increase in volatility the price of both the call and put options will increase by approximately $19.64.

A higher Vega value is (typical) of options that are at the money or close to it-as in this case where the stock price and strike price are both 100. At-the-money options will be more sensitive to changes in volatility because the probability of finishing in the money is heavily influenced by volatility. As time to maturity decreases/or the option moves deeper in or out of the money Vega will tends to decline.

This sensitivity measure is particularly useful in risk management and option trading where Vega will help us gauge the impact of market volatility on option portfolios. High Vega values will imply that an option’s price will fluctuate more significantly in response to volatility changes which makes Vega an essential metric for strategies involving volatility exposure such as like the volatility trading and hedging (Hull, 2018).



 Summary of Step 1

The Black-Scholes model has provided us with valuable insights into the dynamics of the European call and put options:

**Call Option-** We have calculated price and Delta of the call option which suggest that it is at the money. This moderate Delta ( 0.57) has indicated to us that the call option's price will increase moderately with an increases in the stock price.

We have seen that the Vega value will shows a moderate sensitivity to volatility and this will mean that the call option's price will benefit from increases in volatility(although this effect is not as pronounced as it would be for a deeply in-the-money option).

**Put Option:-** Here too the put option's price and Delta indicate an at-the-money status and not deeply "out of the money." This moderate negative Delta (-0.43) reflects that the put option’s price will decrease moderately as the stock price increases. Here the Vega will shows a similar moderate sensitivity to volatility which is indicating that the put option will gain value as volatility rises and this is  making it useful for hedging against potential downside risks in volatile markets.

The Black-Scholes model is effective here because it assumes constant volatility and a risk-free rate and we know that this aligns well with European options that lack early exercise features. This model application provides us with a clear understanding of price sensitivities (Delta and Vega) and potential hedging strategies, especially for at-the-money options (Hull, 2018; McDonald, 2013).

##Team Member B

**Q5** Choosing the number of steps

For a period of 3 months with daily time-steps, the number of steps would be approximately 63 (considering an average of 21 trading days per month).

However, the number of Monte Carlo paths we choose for the simulation affects the reliability of our estimates. As a starting point, let's consider 100,000 paths.
Here, in implementing Step 1 with Monte Carlo simulations for European options, we assume 100,000 simulations for accuracy.

In [ ]:
import numpy as np

# Parameters based on the GWP1 for European options
S0 = 100  # Initial stock price
K = 100   # Strike price
r = 0.05  # Risk-free interest rate
T = 0.25  # Time to maturity in years (3 months)
sigma = 0.2  # Volatility (20%)
days = int(T * 252)  # Daily steps in the simulation for the duration of T
num_simulations = 100000  # Number of Monte Carlo simulations for high accuracy

# Monte Carlo simulation for European option pricing using GBM
np.random.seed(0)
dt = 1 / 252  # Daily time step

# Generate random price paths
price_paths = np.zeros((num_simulations, days))
price_paths[:, 0] = S0

for t in range(1, days):
    # Generate random daily returns based on GBM
    price_paths[:, t] = price_paths[:, t - 1] * np.exp((r - 0.5 * sigma ** 2) * dt + sigma * np.sqrt(dt) * np.random.randn(num_simulations))

# Calculate the payoffs for call and put options at maturity
call_payoffs = np.maximum(price_paths[:, -1] - K, 0)
put_payoffs = np.maximum(K - price_paths[:, -1], 0)

# Discount the payoffs back to present value and average to get option prices
european_call_price_mc = np.exp(-r * T) * np.mean(call_payoffs)
european_put_price_mc = np.exp(-r * T) * np.mean(put_payoffs)

european_call_price_mc, european_put_price_mc


(4.578411635344254, 3.3422485384153005)

The Monte Carlo simulation for European options yielded the following prices:

- European Call Price: 4.58
- European Put Price: 3.34

**5(b) Overall process description**

Initialization: We set our parameters (stock price S_0, risk-free rate r, volatility σ, maturity T T, number of steps, and number of paths). Simulation: For each path:

first, Initialize S_0
second, For each step, simulate the stock price using the discretized GBM formula.
lastly, we calculate the European call and put option payoff at maturity. Aggregation: Average the payoffs across all paths and discount it back to present value using the risk-free rate. Result: The discounted average gives us the option prices.

**Q6**

Next, we calculate Delta by adjusting the initial stock price slightly and observing the impact on the option prices.



In [ ]:
# Small change in stock price for Delta calculation
delta_S = 1  # Small adjustment to initial stock price
S_up = S0 + delta_S  # Slight increase in S0
S_down = S0 - delta_S  # Slight decrease in S0

# Simulate paths for S_up and S_down to calculate Delta for call and put options
price_paths_up = np.zeros((num_simulations, days))
price_paths_down = np.zeros((num_simulations, days))
price_paths_up[:, 0] = S_up
price_paths_down[:, 0] = S_down

for t in range(1, days):
    # Simulate paths for adjusted initial stock prices
    price_paths_up[:, t] = price_paths_up[:, t - 1] * np.exp((r - 0.5 * sigma ** 2) * dt + sigma * np.sqrt(dt) * np.random.randn(num_simulations))
    price_paths_down[:, t] = price_paths_down[:, t - 1] * np.exp((r - 0.5 * sigma ** 2) * dt + sigma * np.sqrt(dt) * np.random.randn(num_simulations))

# Calculate option payoffs for adjusted initial stock prices
call_payoffs_up = np.maximum(price_paths_up[:, -1] - K, 0)
put_payoffs_up = np.maximum(K - price_paths_up[:, -1], 0)
call_payoffs_down = np.maximum(price_paths_down[:, -1] - K, 0)
put_payoffs_down = np.maximum(K - price_paths_down[:, -1], 0)

# Discounted option prices for adjusted stock prices
european_call_price_up = np.exp(-r * T) * np.mean(call_payoffs_up)
european_put_price_up = np.exp(-r * T) * np.mean(put_payoffs_up)
european_call_price_down = np.exp(-r * T) * np.mean(call_payoffs_down)
european_put_price_down = np.exp(-r * T) * np.mean(put_payoffs_down)

# Calculate Delta for call and put options
delta_call = (european_call_price_up - european_call_price_down) / (2 * delta_S)
delta_put = (european_put_price_up - european_put_price_down) / (2 * delta_S)

delta_call, delta_put


(0.5805366616395653, -0.442713723992463)

The calculated Delta values for the European options are:

- Delta for European Call: 0.581
- Delta for European Put: -0.443

6(a) How do they compare

Delta for a call option usually lies between 0 and 1 for European options without dividends. For puts, it is between -1 and 0.

6(b) Comments on differences

Delta for Call Option: Our output shows that the Delta of a call option is positive. This implies that as the stock price increases, the call option's price will also increase. This makes sense: if you have the right to buy a stock at a set price and the stock's market price goes up, the value of that right (i.e., the call option) should also go up.

Delta for Put Option: The Delta of a put option is negative, indicating that as the stock price increases, the put option's price will decrease. Again, this is intuitive: if you have the right to sell a stock at a set price and the stock's market price goes up, the value of that right (i.e., the put option) should decrease since it becomes less likely that the put option will be exercised.

Delta as a Proxy: Delta can also be seen as a proxy for the probability that the option will be in-the-money at expiration. For instance, a Delta of 0.58 for a call option can be roughly interpreted as a 58% chance that the option will expire in-the-money

In summary, the signs of the Deltas for call and put options reflect the directional exposure of the option to the underlying asset. A positive Delta for the call means it benefits from rises in the stock, while a negative Delta for the put means it benefits from falls in the stock.

**Q7**
Now, we proceed with the Vega calculation by increasing the volatility and observing the change in option prices.


In [ ]:
# Increase in volatility for Vega calculation
sigma_new = 0.25  # New volatility (25%)

# Generate new price paths with increased volatility
price_paths_sigma_new = np.zeros((num_simulations, days))
price_paths_sigma_new[:, 0] = S0

for t in range(1, days):
    # Simulate paths with increased volatility
    price_paths_sigma_new[:, t] = price_paths_sigma_new[:, t - 1] * np.exp((r - 0.5 * sigma_new ** 2) * dt + sigma_new * np.sqrt(dt) * np.random.randn(num_simulations))

# Calculate option payoffs for increased volatility
call_payoffs_sigma_new = np.maximum(price_paths_sigma_new[:, -1] - K, 0)
put_payoffs_sigma_new = np.maximum(K - price_paths_sigma_new[:, -1], 0)

# Discounted option prices for new volatility
european_call_price_sigma_new = np.exp(-r * T) * np.mean(call_payoffs_sigma_new)
european_put_price_sigma_new = np.exp(-r * T) * np.mean(put_payoffs_sigma_new)

# Calculate Vega for call and put options
vega_call = (european_call_price_sigma_new - european_call_price_mc) / (sigma_new - sigma)
vega_put = (european_put_price_sigma_new - european_put_price_mc) / (sigma_new - sigma)

vega_call, vega_put


(19.176531150129005, 20.057752875940484)

The calculated Vega values for the European options are:

- Vega for European Call: 19.18
- Vega for European Put: 20.06

**7(a) How do prices change with respect to volatility change**

An increase in volatility leads to a rise in option prices. This is because higher volatility enhances the likelihood of the stock moving favorably for the option holder. Consequently, a 5% volatility increase would boost both call and put option prices.

**7(b) Potential differential impact of this change on the options**

While volatility boosts both call and put option prices, the magnitude of increase varies based on their moneyness.
For at-the-money (ATM) options, like ours, call and put options exhibit similar Vega values, increasing relatively equally with volatility.
However, for deep in-the-money (ITM) or out-of-the-money (OTM) options, Vega differs between calls and puts. Typically, ATM options have the highest Vega, decreasing as you move away from the ATM point.

## Step 2: Monte Carlo Simulation with Geometric Brownian Motion for American Call Option

In this step, we"ll apply the  Monte Carlo simulation methods with a Geometric Brownian Motion (GBM) process to simulate daily prices for an American Call option. The Monte Carlo method allows for flexibility in pricing derivatives moreso for the American options where early exercise can be beneficial. We"ll use the input data from our GWP1 to calculate the option price, delta, and vega.

**Parameters from GWP1**

- Initial Stock Price ($S_0$): 100
- Strike Price ($K$): 100
- Risk-Free Rate ($r$): 0.05%
- Time to Maturity ($T$): 0.25 years (3 months)
- Volatility ($\sigma$): 0.2%
- Number of Simulations ($N$): 10,000
- Steps per Simulation ($M$): 63 steps (these are daily steps over a 3-month period)

These parameters will provide the foundation for modeling the price dynamics of the underlying asset which will follows a stochastic process as per the GBM model.

### Monte Carlo Simulation for American Call Option Pricing

To price the American Call option- we will generate daily price paths and determine the payoff at each step.
American options can/may be exercised at any time and we will apply a backward induction approach to check for optimal exercise points within each path.

In [ ]:
import numpy as np

# Parameters for the American Call Option
S0 = 100         # Initial stock price
K = 100          # Strike price
r = 0.05         # Risk-free interest rate
T = 0.25         # Time to maturity (3 months in years)
sigma = 0.2      # Volatility
N = 10000        # Number of simulations
M = 63           # Steps (daily over 3 months)

# Time increment
dt = T / M
discount_factor = np.exp(-r * dt)

# Generate GBM paths for stock prices
np.random.seed(42)  # For reproducibility
stock_paths = np.zeros((N, M+1))
stock_paths[:, 0] = S0

for t in range(1, M+1):
    z = np.random.standard_normal(N)
    stock_paths[:, t] = stock_paths[:, t-1] * np.exp((r - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * z)

# Initialize option values at maturity
option_values = np.maximum(stock_paths[:, -1] - K, 0)

# Backward induction for early exercise
for t in range(M-1, 0, -1):
    # Identify in-the-money paths
    in_the_money = stock_paths[:, t] > K
    # Calculate immediate exercise payoff
    immediate_exercise = np.maximum(stock_paths[:, t] - K, 0)
    # Perform regression to estimate the continuation value
    if in_the_money.any():
        X = stock_paths[in_the_money, t]
        Y = option_values[in_the_money] * discount_factor
        # Polynomial regression on stock price (X) and discounted future payoff (Y)
        reg = np.polyfit(X, Y, 2)
        continuation_value = np.polyval(reg, X)
        # Update option values with the maximum of continuation and immediate exercise
        option_values[in_the_money] = np.where(immediate_exercise[in_the_money] > continuation_value,
                                               immediate_exercise[in_the_money],
                                               option_values[in_the_money] * discount_factor)
    else:
        # Discount future payoff if no in-the-money paths
        option_values = option_values * discount_factor

# Discount option value back to present
american_call_price = np.mean(option_values) * np.exp(-r * T)

print(f"American Call Option Price: {american_call_price:.2f}")


American Call Option Price: 4.39


###Delta Calculation

We already know that -Delta represents the sensitivity of the option price to small changes in the underlying asset's price.

For the American Call option-delta is approximated by calculating the difference in simulated option prices for small changes in the initial stock price:

In [ ]:
# Delta approximation for American Call
epsilon = 1e-4
S0_up = S0 * (1 + epsilon)
S0_down = S0 * (1 - epsilon)

# Re-run simulations with adjusted initial stock prices for delta calculation
stock_paths_up = np.zeros((N, M + 1))
stock_paths_down = np.zeros((N, M + 1))
stock_paths_up[:, 0] = S0_up
stock_paths_down[:, 0] = S0_down

for t in range(1, M + 1):
    z = np.random.standard_normal(N)
    stock_paths_up[:, t] = stock_paths_up[:, t - 1] * np.exp((r - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * z)
    stock_paths_down[:, t] = stock_paths_down[:, t - 1] * np.exp((r - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * z)

# Calculate payoffs for delta scenarios
payoff_up = np.maximum(stock_paths_up[:, -1] - K, 0)
payoff_down = np.maximum(stock_paths_down[:, -1] - K, 0)

# Discounted mean payoffs
american_call_price_up = discount_factor * np.mean(payoff_up)
american_call_price_down = discount_factor * np.mean(payoff_down)

# Delta calculation
delta_american_call = (american_call_price_up - american_call_price_down) / (S0_up - S0_down)
print(f"Delta for American Call Option: {delta_american_call:.4f}")


Delta for American Call Option: 0.5719


In [ ]:
# Delta approximation for American Call
epsilon = 1e-4
S0_up = S0 * (1 + epsilon)
S0_down = S0 * (1 - epsilon)

# Re-run simulations with adjusted initial stock prices for delta calculation
stock_paths_up = np.zeros((N, M + 1))
stock_paths_down = np.zeros((N, M + 1))
stock_paths_up[:, 0] = S0_up
stock_paths_down[:, 0] = S0_down

for t in range(1, M + 1):
    z = np.random.standard_normal(N)
    stock_paths_up[:, t] = stock_paths_up[:, t - 1] * np.exp((r - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * z)
    stock_paths_down[:, t] = stock_paths_down[:, t - 1] * np.exp((r - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * z)

# Calculate payoffs for delta scenarios
payoff_up = np.maximum(stock_paths_up[:, -1] - K, 0)
payoff_down = np.maximum(stock_paths_down[:, -1] - K, 0)

# Discounted mean payoffs
american_call_price_up = discount_factor * np.mean(payoff_up)
american_call_price_down = discount_factor * np.mean(payoff_down)

# Delta calculation
delta_american_call = (american_call_price_up - american_call_price_down) / (S0_up - S0_down)
print(f"Delta for American Call Option: {delta_american_call:.4f}")


Delta for American Call Option: 0.5751


###Vega Calculation
Vega-- measures the sensitivity of the option price to changes in volatility. To estimate vega we will rerun the Monte Carlo simulations with a small adjustment in volatility and compute the difference in option prices.

In [ ]:
# Vega approximation for American Call
sigma_up = sigma * (1 + epsilon)
sigma_down = sigma * (1 - epsilon)

# Re-run simulations with adjusted volatility for vega calculation
stock_paths_sigma_up = np.zeros((N, M + 1))
stock_paths_sigma_down = np.zeros((N, M + 1))
stock_paths_sigma_up[:, 0] = S0
stock_paths_sigma_down[:, 0] = S0

for t in range(1, M + 1):
    z = np.random.standard_normal(N)
    stock_paths_sigma_up[:, t] = stock_paths_sigma_up[:, t - 1] * np.exp((r - 0.5 * sigma_up**2) * dt + sigma_up * np.sqrt(dt) * z)
    stock_paths_sigma_down[:, t] = stock_paths_sigma_down[:, t - 1] * np.exp((r - 0.5 * sigma_down**2) * dt + sigma_down * np.sqrt(dt) * z)

# Calculate payoffs for vega scenarios
payoff_sigma_up = np.maximum(stock_paths_sigma_up[:, -1] - K, 0)
payoff_sigma_down = np.maximum(stock_paths_sigma_down[:, -1] - K, 0)

# Discounted mean payoffs
american_call_price_sigma_up = discount_factor * np.mean(payoff_sigma_up)
american_call_price_sigma_down = discount_factor * np.mean(payoff_sigma_down)

# Vega calculation
vega_american_call = (american_call_price_sigma_up - american_call_price_sigma_down) / (sigma_up - sigma_down)
print(f"Vega for American Call Option: {vega_american_call:.4f}")


Vega for American Call Option: 19.4905


In [ ]:
import numpy as np

# Parameters
S0 = 100         # Initial stock price
K = 100          # Strike price
r = 0.05         # Risk-free interest rate
T = 0.25         # Time to maturity (3 months)
sigma = 0.2      # Initial volatility
N = 10000        # Number of simulations
M = 63           # Steps (daily over 3 months)
epsilon = 1e-2   # Small change for Vega calculation

# Time increment
dt = T / M
discount_factor = np.exp(-r * dt)

# Function to simulate stock paths and compute American call price with early exercise
def simulate_american_call(S0, sigma):
    stock_paths = np.zeros((N, M + 1))
    stock_paths[:, 0] = S0

    for t in range(1, M + 1):
        z = np.random.standard_normal(N)
        stock_paths[:, t] = stock_paths[:, t - 1] * np.exp((r - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * z)

    # Initialize option values at maturity
    option_values = np.maximum(stock_paths[:, -1] - K, 0)

    # Backward induction for early exercise
    for t in range(M - 1, 0, -1):
        in_the_money = stock_paths[:, t] > K
        immediate_exercise = np.maximum(stock_paths[:, t] - K, 0)

        if in_the_money.any():
            X = stock_paths[in_the_money, t]
            Y = option_values[in_the_money] * discount_factor
            reg = np.polyfit(X, Y, 2)
            continuation_value = np.polyval(reg, X)

            option_values[in_the_money] = np.where(immediate_exercise[in_the_money] > continuation_value,
                                                   immediate_exercise[in_the_money],
                                                   option_values[in_the_money] * discount_factor)
        else:
            option_values = option_values * discount_factor

    # Return discounted present value of the option
    return np.mean(option_values) * np.exp(-r * T)

# Calculate Vega using absolute adjustment
np.random.seed(42)  # Consistent paths for comparison
american_call_price_sigma_up = simulate_american_call(S0, sigma + epsilon)
np.random.seed(42)  # Reset seed for identical paths
american_call_price_sigma_down = simulate_american_call(S0, sigma - epsilon)

# Vega calculation
vega_american_call = (american_call_price_sigma_up - american_call_price_sigma_down) / (2 * epsilon)
print(f"Vega for American Call Option: {vega_american_call:.4f}")


Vega for American Call Option: 19.0601


### Results

we have obtained the following values for the American call option-after running the Monte Carlo simulations,
- **American Call Option Price**= 4.39
- **Delta for American Call Option**= 0.5719
- **Vega for American Call Option**= 19.4905

### Analysis of our  Results

**American Call Option Price (4.39)-**

This value will aligns with expectations for an at-the-money call option with moderate sensitivity to price changes given that the current stock price is close to the strike price. The price will reflects the flexibility of early exercise in American options, which adds a small premium compared to a European option with the same parameters.

**Delta (0.5719)-**

The Delta value of 0.5719 will indicates moderate sensitivity to the underlying stock price. This is (typical) for an at-the-money option-where there is a balanced probability of the option ending in or out of the money. This Delta will suggests that a 1unit increase in the underlying asset’s price will lead to a 0.5719-unit increase in the option price.

**Vega (19.4905)-**

The Vega value of 19.4905-indicates a high sensitivity to changes in volatility. This will mean that a 1% increase in volatility would increase the option price by about 19.49 units. High Vega values are for at-the-money options beacuse their value is more sensitive to volatility changes compared to deep in-the-money or out-of-the-money options. This sensitivity is valuable for managing exposure to market volatility.

### Summary of Step 2 Analysis

- **American Call Option Price**- Our Monte Carlo estimate of 4.39 will provides us  a reliable value for an at-the-money American call option with the added flexibility of early exercise.
- **Delta**-The Delta of 0.5719-suggests moderate sensitivity to the underlying stock price as we expected for an at-the-money option.
- **Vega**-A Vega of 19.4905-implies substantial sensitivity to changes in volatility-This'sa characteristic of options close to the strike price where volatility plays a significant role in their valuation.


**Team member B**

we price an American Put option using Monte Carlo methods with a daily time-step under a Geometric Brownian Motion (GBM) process. To ensure accurate American-style option pricing (where early exercise is allowed), we'll implement the Longstaff-Schwartz algorithm, which is designed for optimal stopping in Monte Carlo simulations.

(Q5): Calculate the American Put option price by simulating paths under the GBM process and applying the Longstaff-Schwartz method to decide on early exercise.

In [ ]:
# Parameters specific to the American Put option pricing
put_payoffs_american = np.zeros(num_simulations)

# Initialize price paths for American Put option
price_paths_american = np.zeros((num_simulations, days))
price_paths_american[:, 0] = S0

# Simulate paths under GBM for the American Put option
for t in range(1, days):
    price_paths_american[:, t] = price_paths_american[:, t - 1] * np.exp((r - 0.5 * sigma ** 2) * dt + sigma * np.sqrt(dt) * np.random.randn(num_simulations))

# Initialize payoffs at maturity
put_payoffs_american = np.maximum(K - price_paths_american[:, -1], 0)

# Backward induction for early exercise using Longstaff-Schwartz
for t in range(days - 2, 0, -1):
    in_the_money = price_paths_american[:, t] < K  # In-the-money paths
    payoff_if_exercised = np.maximum(K - price_paths_american[in_the_money, t], 0)
    continuation_values = np.exp(-r * dt) * put_payoffs_american[in_the_money]

    # Fit regression to estimate continuation values
    if len(payoff_if_exercised) > 0:  # Only if there are paths in the money
        regression = np.polyfit(price_paths_american[in_the_money, t], continuation_values, 2)
        continuation_estimate = np.polyval(regression, price_paths_american[in_the_money, t])

        # Exercise if the payoff is greater than the continuation value
        exercise = payoff_if_exercised > continuation_estimate
        put_payoffs_american[in_the_money] = np.where(exercise, payoff_if_exercised, continuation_values)

# Discount the first time step to obtain the American Put price
american_put_price_mc = np.exp(-r * dt) * np.mean(put_payoffs_american)
american_put_price_mc


3.450232587772447

The calculated price for the American Put option using the Monte Carlo and Longstaff-Schwartz method is 3.45.

Next, we estimate Delta by adjusting the initial stock price and observing the impact on the American put option price.

In [ ]:
# Adjust initial stock prices for Delta calculation
price_paths_american_up = np.zeros((num_simulations, days))
price_paths_american_down = np.zeros((num_simulations, days))
price_paths_american_up[:, 0] = S_up
price_paths_american_down[:, 0] = S_down

# Simulate paths for adjusted initial stock prices
for t in range(1, days):
    price_paths_american_up[:, t] = price_paths_american_up[:, t - 1] * np.exp((r - 0.5 * sigma ** 2) * dt + sigma * np.sqrt(dt) * np.random.randn(num_simulations))
    price_paths_american_down[:, t] = price_paths_american_down[:, t - 1] * np.exp((r - 0.5 * sigma ** 2) * dt + sigma * np.sqrt(dt) * np.random.randn(num_simulations))

# Initialize payoffs for adjusted stock prices at maturity
put_payoffs_american_up = np.maximum(K - price_paths_american_up[:, -1], 0)
put_payoffs_american_down = np.maximum(K - price_paths_american_down[:, -1], 0)

# Backward induction for adjusted stock prices using Longstaff-Schwartz
for t in range(days - 2, 0, -1):
    in_the_money_up = price_paths_american_up[:, t] < K
    payoff_if_exercised_up = np.maximum(K - price_paths_american_up[in_the_money_up, t], 0)
    continuation_values_up = np.exp(-r * dt) * put_payoffs_american_up[in_the_money_up]

    if len(payoff_if_exercised_up) > 0:
        regression_up = np.polyfit(price_paths_american_up[in_the_money_up, t], continuation_values_up, 2)
        continuation_estimate_up = np.polyval(regression_up, price_paths_american_up[in_the_money_up, t])
        exercise_up = payoff_if_exercised_up > continuation_estimate_up
        put_payoffs_american_up[in_the_money_up] = np.where(exercise_up, payoff_if_exercised_up, continuation_values_up)

    in_the_money_down = price_paths_american_down[:, t] < K
    payoff_if_exercised_down = np.maximum(K - price_paths_american_down[in_the_money_down, t], 0)
    continuation_values_down = np.exp(-r * dt) * put_payoffs_american_down[in_the_money_down]

    if len(payoff_if_exercised_down) > 0:
        regression_down = np.polyfit(price_paths_american_down[in_the_money_down, t], continuation_values_down, 2)
        continuation_estimate_down = np.polyval(regression_down, price_paths_american_down[in_the_money_down, t])
        exercise_down = payoff_if_exercised_down > continuation_estimate_down
        put_payoffs_american_down[in_the_money_down] = np.where(exercise_down, payoff_if_exercised_down, continuation_values_down)

# Discount the first time step to obtain American Put prices for adjusted initial stock prices
american_put_price_up = np.exp(-r * dt) * np.mean(put_payoffs_american_up)
american_put_price_down = np.exp(-r * dt) * np.mean(put_payoffs_american_down)

# Calculate Delta for the American Put option
delta_american_put = (american_put_price_up - american_put_price_down) / (2 * delta_S)
delta_american_put


-0.44060924498660703

The calculated Delta for the American Put option is -0.44.

**Comment on Delta:**

A negative Delta (typically between 0 and -1 for equity puts) implies that the option price will drop when the stock price increases, and vice versa. This makes sense for put options because they become more valuable as the stock price decreases. A positive Delta ( rare and unusual for put options) would imply the option price increases when the stock price increases, which is not the usual behavior for put options. For an American put option, you would expect the Delta to be negative. The reason is that as the stock price increases, the intrinsic value of the put (the benefit from exercising early) diminishes, leading to a decrease in its overall value. on the otherhand, if the stock price drops, the potential for profit from the put option increases, making it more valuable.

Now for **Q7** we recalculate the option price with a shocked volatility parameter to estimate Vega.

In [ ]:
# Initialize paths for increased volatility for Vega calculation
price_paths_american_sigma_new = np.zeros((num_simulations, days))
price_paths_american_sigma_new[:, 0] = S0

# Simulate paths under increased volatility
for t in range(1, days):
    price_paths_american_sigma_new[:, t] = price_paths_american_sigma_new[:, t - 1] * np.exp((r - 0.5 * sigma_new ** 2) * dt + sigma_new * np.sqrt(dt) * np.random.randn(num_simulations))

# Initialize payoffs at maturity for increased volatility
put_payoffs_american_sigma_new = np.maximum(K - price_paths_american_sigma_new[:, -1], 0)

# Backward induction for American Put option with increased volatility using Longstaff-Schwartz
for t in range(days - 2, 0, -1):
    in_the_money_sigma_new = price_paths_american_sigma_new[:, t] < K
    payoff_if_exercised_sigma_new = np.maximum(K - price_paths_american_sigma_new[in_the_money_sigma_new, t], 0)
    continuation_values_sigma_new = np.exp(-r * dt) * put_payoffs_american_sigma_new[in_the_money_sigma_new]

    if len(payoff_if_exercised_sigma_new) > 0:
        regression_sigma_new = np.polyfit(price_paths_american_sigma_new[in_the_money_sigma_new, t], continuation_values_sigma_new, 2)
        continuation_estimate_sigma_new = np.polyval(regression_sigma_new, price_paths_american_sigma_new[in_the_money_sigma_new, t])
        exercise_sigma_new = payoff_if_exercised_sigma_new > continuation_estimate_sigma_new
        put_payoffs_american_sigma_new[in_the_money_sigma_new] = np.where(exercise_sigma_new, payoff_if_exercised_sigma_new, continuation_values_sigma_new)

# Discount the first time step to obtain American Put price for increased volatility
american_put_price_sigma_new = np.exp(-r * dt) * np.mean(put_payoffs_american_sigma_new)

# Calculate Vega for the American Put option
vega_american_put = (american_put_price_sigma_new - american_put_price_mc) / (sigma_new - sigma)
vega_american_put


19.110512106215026

The calculated Vega for the American Put option is 19.11.

**Comments on the impact of volatility change on American put**

The American put option's early exercise feature typically makes it more valuable than its European counterpart, especially for higher volatility scenarios. Higher volatility increases the likelihood of the option becoming deep in-the-money prior to maturity, offering the holder optimal exercise opportunities. As volatility expands the range of potential stock prices, it enhances the possibility of beneficial downside price movements for American put option holders. This heightened possibility is shown in the option price increase accompanying rising volatility.

### Step 3: Hedging Under Black-Scholes for European Options and Pricing Different Exotic Instruments

This step 3 here involves using the Black-Scholes model to price European options with specific moneyness levels and calculate their delta. Subsequently, we will then construct two portfolios based on these options and determine the appropriate hedging strategies.

The Black-Scholes -This's is a model used to calculate the prices of European call and put options along with their sensitivities like Delta.

The call and put option prices and Delta under the Black-Scholes model (Black & Scholes, 1973; Hull, 2018) are defined as below.

**for a European Call Option**

the Black-Scholes formula for the option price for an European call option on a non-dividend-paying stock=

$$
\text{Call Price} = S_0 \cdot N(d_1) - K \cdot e^{-rT} \cdot N(d_2)
$$

Where:

- $S_0$: Current stock price
- $K$: Strike price of the option
- $r$: Risk-free interest rate
- $T$: Time to expiration (in years)
- $\sigma$: Volatility of the stock price (annualized)
- $N(d)$: Cumulative distribution function of the standard normal distribution

$d_1$ and $d_2$ are intermediate calculations given by:

$$
d_1 = \frac{\ln \left( \frac{S_0}{K} \right) + \left( r + \frac{\sigma^2}{2} \right) T}{\sigma \sqrt{T}}
$$

$$
d_2 = d_1 - \sigma \sqrt{T}
$$


**For a European put option, the formula is:**

$$
\text{Put Price} = K \cdot e^{-rT} \cdot N(-d_2) - S_0 \cdot N(-d_1)
$$

The parameters $S_0$, $K$, $r$, $T$, $\sigma$, and $d_1$ and $d_2$ are the same as those defined above.


Again we know that the Delta of an option measures the sensitivity of the option price to changes in the underlying stock price (Hull, 2018).

For a call option, Delta ($\Delta_{\text{call}}$) is:

$$
\Delta_{\text{call}} = N(d_1)
$$

For a put option, Delta ($\Delta_{\text{put}}$) is:

$$
\Delta_{\text{put}} = N(d_1) - 1
$$

Where:

- $N(d_1)$: The cumulative probability up to $d_1$ under the standard normal distribution, representing the likelihood of the option expiring in the money.

>>> ### Summary of Formulas

- **Call Price**:
  $$
  \text{Call Price} = S_0 \cdot N(d_1) - K \cdot e^{-rT} \cdot N(d_2)
  $$

 \

- **Put Price**:
  $$
  \text{Put Price} = K \cdot e^{-rT} \cdot N(-d_2) - S_0 \cdot N(-d_1)
  $$

  \

- **Call Delta**:
  $$
  \Delta_{\text{call}} = N(d_1)
  $$

  \


- **Put Delta**:
  $$
  \Delta_{\text{put}} = N(d_1) - 1
  $$

We derive these formulas from the Black-Scholes model-which assumes constant volatility, a known risk-free rate and continuous time and which makes it suitable for pricing European options that cannot be exercised early (Black & Scholes, 1973; Hull, 2018).


### Part 1: Pricing European Call and Put Options with Different Moneyness Levels

In this part we will price a European call option with 110% moneyness and a European put option with 95% moneyness and both with a maturity of 3 months.
Lets apply The Black-Scholes formula to calculate the option prices and delta as->



**Input Parameters:**

- $S_0$: Current stock price
- $K$: Strike price (adjusted for moneyness)
- $T$: Time to maturity (3 months, or 0.25 years)
- $r$: Risk-free interest rate
- $\sigma$: Volatility




**Pricing and Delta Calculation**

In [ ]:
import numpy as np
from scipy.stats import norm

# Parameters
S0 = 100           # Initial stock price
r = 0.05           # Risk-free rate
sigma = 0.2        # Volatility
T = 0.25           # Time to maturity (3 months)

# 1. European Call Option with 110% Moneyness
K_call = S0 * 1.1  # Strike price for the call option (110% moneyness)
d1_call = (np.log(S0 / K_call) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
d2_call = d1_call - sigma * np.sqrt(T)
call_price = S0 * norm.cdf(d1_call) - K_call * np.exp(-r * T) * norm.cdf(d2_call)
delta_call = norm.cdf(d1_call)

# 2. European Put Option with 95% Moneyness
K_put = S0 * 0.95  # Strike price for the put option (95% moneyness)
d1_put = (np.log(S0 / K_put) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
d2_put = d1_put - sigma * np.sqrt(T)
put_price = K_put * np.exp(-r * T) * norm.cdf(-d2_put) - S0 * norm.cdf(-d1_put)
delta_put = norm.cdf(d1_put) - 1  # For a put option, delta = N(d1) - 1

# Display results
print(f"European Call Option Price with 110% Moneyness: {call_price:.2f}")
print(f"Delta for European Call Option with 110% Moneyness: {delta_call:.4f}\n")

print(f"European Put Option Price with 95% Moneyness: {put_price:.2f}")
print(f"Delta for European Put Option with 95% Moneyness: {delta_put:.4f}\n")


European Call Option Price with 110% Moneyness: 1.19
Delta for European Call Option with 110% Moneyness: 0.2183

European Put Option Price with 95% Moneyness: 1.53
Delta for European Put Option with 95% Moneyness: -0.2457



In [ ]:
import numpy as np
from scipy.stats import norm

# Parameters
S0 = 100           # Initial stock price
r = 0.05           # Risk-free interest rate
sigma = 0.2        # Volatility
T = 0.25           # Time to maturity (3 months)

# 1. European Call Option with 110% Moneyness
K_call = S0 * 1.1  # Strike price for the call option (110% moneyness)
d1_call = (np.log(S0 / K_call) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
d2_call = d1_call - sigma * np.sqrt(T)
call_price = S0 * norm.cdf(d1_call) - K_call * np.exp(-r * T) * norm.cdf(d2_call)
delta_call = norm.cdf(d1_call)

# 2. European Put Option with 95% Moneyness
K_put = S0 * 0.95  # Strike price for the put option (95% moneyness)
d1_put = (np.log(S0 / K_put) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
d2_put = d1_put - sigma * np.sqrt(T)
put_price = K_put * np.exp(-r * T) * norm.cdf(-d2_put) - S0 * norm.cdf(-d1_put)
delta_put = norm.cdf(d1_put) - 1  # For a put option, delta = N(d1) - 1

# Display results
print(f"European Call Option Price with 110% Moneyness: {call_price:.2f}")
print(f"Delta for European Call Option with 110% Moneyness: {delta_call:.4f}\n")

print(f"European Put Option Price with 95% Moneyness: {put_price:.2f}")
print(f"Delta for European Put Option with 95% Moneyness: {delta_put:.4f}\n")


European Call Option Price with 110% Moneyness: 1.19
Delta for European Call Option with 110% Moneyness: 0.2183

European Put Option Price with 95% Moneyness: 1.53
Delta for European Put Option with 95% Moneyness: -0.2457



### Our Results and Interpreting them:

**European Call Option with 110% Moneyness:**
- Price= 1.19
- Delta=0.2183

**European Put Option with 95% Moneyness:**
- Price= 1.52
- Delta= -0.2457

Our output/results confirm that out-of-the-money call options have lower prices and deltas, while in-the-money put options reflect higher values. These results are consistent with theoretical expectations in options pricing (Hull, 2018; Black & Scholes, 1973).

### Part 2: Building and Hedging Portfolios Based on Moneyness Levels

we willl construct and hedge two portfolios using our  calculated option prices and deltas,

- **Portfolio 1**: here we'll be Buying one European call with 110% moneyness and one European put with 95% moneyness.
- **Portfolio 2**: while here we will be Buying one European call with 110% moneyness and selling one European put with 95% moneyness.

#### Portfolio 1: Buying Both Call and Put Options (Straddle Position)

This portfolio will combines a long position in both the call and put options forming a straddle that typically profits from volatility rather than price directionality.

$$
\Delta_{\text{Portfolio 1}} = \Delta_{\text{Call}} + \Delta_{\text{Put}} = 0.2183 + (-0.2457) = -0.0275
$$

To hedge this portfolio-we will aim to make it delta-neutral by purchasing 0.0275 units of the underlying asset. That is-the hedge strategy would be to buy 0.0275 units of the underlying asset to make the portfolio delta-neutral.




### Part 2: Constructing and Hedging Portfolios Based on Moneyness Levels

Hre we'l examine two portfolios-we have constructed using European call and put options with different moneyness levels.

- **Portfolio 1**= Comprised of one long European call option with 110% moneyness and one long European put option with 95% moneyness.
- **Portfolio 2**=Comprised of one long European call option with 110% moneyness and one short European put option with 95% moneyness.

we will calculate the initial portfolio values and aggregate Deltas, enabling us to determine the optimal hedging strategies Using the Black-Scholes model,.

### Results

we have calculated thevalues for each portfolio are as follows-

- **Portfolio 1 (Long Call 110% + Long Put 95%)**:
  - Our Initial Value: 2.72
  - The Aggregate Delta: -0.0274
  - Our  Hedging Strategy: Buy 0.0274 units of the underlying asset

- **Portfolio 2 (Long Call 110% + Short Put 95%)**:
  - Here we have the Initial Value= -0.34
  - TheAggregate Delta= 0.4640
  - The Hedging Strategy=Short-sell 0.4640 units of the underlying asset

### Lets Analysis

**Portfolio 1- Long Call and Long Put**

- **Initial Value**-The initial value of 2.72 represents-the total cost of acquiring both the call and put options at the specified moneyness levels. This positive value will indicates an outlay of capital and is reflecting the cost of purchasing both options.
- **Aggregate Delta (-0.0274)**=The aggregate Delta of -0.0274-indicates that the portfolio has a slight negative exposure to the underlying asset’s price. This will means that the portfolio will gain marginally if the asset price decreases and lose slightly if the asset price increases.
- **Hedging Strategy**=To neutralize this minor negative Delta-we will apply a long hedge by purchasing 0.0274 units of the underlying asset. This position will counteracts the portfolio’s sensitivity to upward movements in the stock price and ensuring a more Delta-neutral stance.

**Portfolio 2: Long Call and Short Put**

- **Initial Value**=The initial value of -0.34 suggests a net credit- this is meaning taht the portfolio generates a premium due to the proceeds from selling the put option (which has a higher price than the call option). This value will signifies that the portfolio structure benefits from the initial inflow of capital.
- **Aggregate Delta (0.4640)**=The positive Delta of 0.4640-indicates a substantial positive sensitivity to changes in the underlying stock price. This willl mean that the portfolio will appreciate if the stock price rises and depreciate if it falls.
- **Hedging Strategy**=To offset the positive Delta exposure-we"ll apply a short hedge by short-selling 0.4640 units of the underlying asset. This action mitigates the portfolio’s sensitivity to the underlying asset’s price, creating a more Delta-neutral position.

### Lets do Interpretation

The construction and analysis of these portfolios has  demonstrated to us the strategic application of options with different moneyness levels to manage directional exposure. In both cases- theDelta hedging is employed to neutralize the portfolios' sensitivity to underlying price movements.

- **Portfolio 1** We have structured it as a balanced position with both call and put options-resulting in a nearly Delta-neutral position. The minor negative Delta will indicate that the portfolio requires only a small hedge in the form of a long position in the underlying asset to achieve a neutral exposure. This is to mean that this structure is suitable for investors seeking a minimal directional bias with the flexibility of benefiting from price movements in either direction.

- **Portfolio 2**-thisby contrast -introduces a short put position, which inversely affects the portfolio’s directional exposure. Thisresultant positive Delta will necessitates a larger short position in the underlying asset to hedge against upward price sensitivity. This structure here is beneficial for traders aiming to collect premiums from the put option sale while maintaining limited exposure to downward price movements through Delta hedging.

In summary, these both portfolios illustrate distinct approaches to managing price sensitivity through options with varied moneyness levels.
Delta hedging-is effectively used in each portfolio to control the directional risk which making these strategies adaptable for different market views and risk preferences.


**Hedging**

In [ ]:
import numpy as np

# Prices and Deltas from Part 1
call_price_110 = 1.19
delta_call_110 = 0.2183
put_price_95 = 1.53
delta_put_95 = -0.2457

# Portfolio 1: Buy one European call (110% moneyness) and buy one European put (95% moneyness)
portfolio1_value = call_price_110 + put_price_95
portfolio_delta_1 = delta_call_110 + delta_put_95
hedging_strategy_1 = "Short-sell" if portfolio_delta_1 > 0 else "Buy"

# Portfolio 2: Buy one European call (110% moneyness) and sell one European put (95% moneyness)
portfolio2_value = call_price_110 - put_price_95
portfolio_delta_2 = delta_call_110 - delta_put_95
hedging_strategy_2 = "Short-sell" if portfolio_delta_2 > 0 else "Buy"

# Display results for Portfolio 1
print("Portfolio 1 (Buying Call 110% + Buying Put 95%):")
print(f"Initial Value: {portfolio1_value:.2f}")
print(f"Delta for Hedging: {portfolio_delta_1:.4f}")
print(f"Hedging Strategy: {hedging_strategy_1} {abs(portfolio_delta_1):.4f} units of the underlying asset\n")

# Display results for Portfolio 2
print("Portfolio 2 (Buying Call 110% + Selling Put 95%):")
print(f"Initial Value: {portfolio2_value:.2f}")
print(f"Delta for Hedging: {portfolio_delta_2:.4f}")
print(f"Hedging Strategy: {hedging_strategy_2} {abs(portfolio_delta_2):.4f} units of the underlying asset\n")


Portfolio 1 (Buying Call 110% + Buying Put 95%):
Initial Value: 2.72
Delta for Hedging: -0.0274
Hedging Strategy: Buy 0.0274 units of the underlying asset

Portfolio 2 (Buying Call 110% + Selling Put 95%):
Initial Value: -0.34
Delta for Hedging: 0.4640
Hedging Strategy: Short-sell 0.4640 units of the underlying asset



### Portfolio 2: Buying Call and Selling Put (Synthetic Long Position)

This portfolio will simulates a synthetic long position with an overall positive delta.

$$
\Delta_{\text{Portfolio 2}} = \Delta_{\text{Call}} - \Delta_{\text{Put}} = 0.2183 - (-0.2457) = 0.4640
$$

The hedge strategy will be to short-sell 0.4640 units of the underlying asset to achieve delta neutrality.


### Analysis and Interpretation

**Portfolio 1: Buying Both Call and Put Options**

 **Delta of Portfolio**: -0.0275
- This portfolio has a slightly negative delta- this indicates a minor sensitivity to price changes in the underlying asset. A small increase in the asset’s price will decrease the portfolio's value while on the other hand a decrease will increase it slightly.
- **Hedging Strategy**:- To neutralize this sensitivity= we will buy 0.0275 units of the underlying asset, making the portfolio delta-neutral.

**Portfolio 2: Buying Call and Selling Put Options**

 **Delta of Portfolio**: 0.4640
-  This portfolio has a positive delta-indicating that it is positively sensitive to changes in the underlying asset's price. if we have an increase in the asset's price then we will increase the portfolio’s value.
- **Hedging Strategy**: To counterbalance this positive delta we will need to short-sell 0.4640 units of the underlying asset and thus we'll be achieving a delta-neutral
position.

### Summary of the Hedging Strategies:

- For **Portfolio 1** (which ahs a near-zero negative delta)- buying a small amount of the underlying asset will balances the portfolio and in turn protecting it from minor fluctuations in the underlying asset's price.

- For **Portfolio 2** (here with a positive delta)- short-selling a portion of the underlying asset will and effectively balance the portfolio and in turns reducing its sensitivity to changes in the underlying price.

These hedging strategies will work aiming to make both portfolios "delta-neutral,"  and this is to mean that their values will be less impacted by small changes in the underlying asset's price.


### Discussion

The calculated deltas and hedging strategies will reflect the behavior of options under the Black-Scholes model and of which is often used for European options coz of its assumptions of continuous hedging and the absence of early exercise (Hull, 2018).

When we adjust the portfolio composition based on delta values, we will create  a delta-neutral strategies that mitigate exposure to the underlying asset's price movements (Black & Scholes, 1973).



### Conclusion

Our  step here validates the pricing and delta calculations for European options with adjusted moneyness levels and demonstrates the application of delta-neutral hedging for two distinct portfolios. The strategies formulated will ensure that each portfolio remains minimally affected by small price changes in the underlying asset.

These findings are aligning with  hedging principles- where portfolio delta values close to zero (Portfolio 1) reflect a neutral stance and while positive delta values (Portfolio 2) require short positions for neutrality. Such and thes strategies are fundamental in managing exposure in options trading and are consistent with standard options pricing and hedging models (Black & Scholes, 1973; Hull, 2018).


**Member B**

Here in Step 3, we will price an Up-and-Out (UAO) Barrier Option using Monte Carlo methods under a Black-Scholes framework. This barrier option is ATM with an up-and-out barrier set at 141, meaning the option becomes worthless if the price of the underlying asset reaches or exceeds this barrier level during the option's life.

Next using the given parameters , we simulate daily daily price paths and check if any price exceeds the barrier level. If so, that path's payoff is set to zero. Otherwise, we compute the option payoff as max(ST−K, 0).

Then **hedging under Black-Scholes** we implement a simple Black-Scholes hedge by calculating Delta for this UAO option.

In [ ]:
# Parameters for the Up-and-Out Barrier Option
S0_UAO = 120  # Initial stock price
K_UAO = 120  # ATM strike price
r_UAO = 0.06  # Risk-free interest rate
sigma_UAO = 0.30  # Volatility (30%)
T_UAO = 8 / 12  # Time to maturity in years (8 months)
days_UAO = int(T_UAO * 252)  # Daily steps for 8 months
barrier_UAO = 141  # Up-and-Out barrier level

# Monte Carlo simulation for Up-and-Out Barrier option pricing
np.random.seed(0)
dt_UAO = 1 / 252  # Daily time step for 252 trading days in a year

# Initialize price paths
price_paths_UAO = np.zeros((num_simulations, days_UAO))
price_paths_UAO[:, 0] = S0_UAO

# Simulate paths under GBM with daily steps
for t in range(1, days_UAO):
    price_paths_UAO[:, t] = price_paths_UAO[:, t - 1] * np.exp((r_UAO - 0.5 * sigma_UAO ** 2) * dt_UAO + sigma_UAO * np.sqrt(dt_UAO) * np.random.randn(num_simulations))

# Check if any path breaches the barrier and calculate payoffs
breached_barrier = (price_paths_UAO >= barrier_UAO).any(axis=1)
payoffs_UAO = np.where(breached_barrier, 0, np.maximum(price_paths_UAO[:, -1] - K_UAO, 0))

# Discount the payoffs to present value
UAO_option_price_mc = np.exp(-r_UAO * T_UAO) * np.mean(payoffs_UAO)
UAO_option_price_mc


0.7088518137656253

The calculated price for the Up-and-Out Barrier option using Monte Carlo simulation is 0.71.

In [ ]:
# Re-import necessary libraries and redefine parameters due to environment reset
import numpy as np

# Parameters for the Up-and-Out Barrier Option
S0_UAO = 120  # Initial stock price
K_UAO = 120  # ATM strike price
r_UAO = 0.06  # Risk-free interest rate
sigma_UAO = 0.30  # Volatility (30%)
T_UAO = 8 / 12  # Time to maturity in years (8 months)
days_UAO = int(T_UAO * 252)  # Daily steps for 8 months
barrier_UAO = 141  # Up-and-Out barrier level
num_simulations = 100000  # Number of simulations for Monte Carlo

# Small change in initial stock price to estimate Delta
delta_S_UAO = 1
S_up_UAO = S0_UAO + delta_S_UAO  # Slightly increased initial price
S_down_UAO = S0_UAO - delta_S_UAO  # Slightly decreased initial price

# Monte Carlo simulation for Delta calculation of Up-and-Out Barrier Option
dt_UAO = 1 / 252  # Daily time step for 252 trading days in a year
np.random.seed(0)

# Initialize price paths for Delta calculation with adjusted initial prices
price_paths_UAO_up = np.zeros((num_simulations, days_UAO))
price_paths_UAO_down = np.zeros((num_simulations, days_UAO))
price_paths_UAO_up[:, 0] = S_up_UAO
price_paths_UAO_down[:, 0] = S_down_UAO

# Simulate paths for both slightly increased and decreased initial stock prices
for t in range(1, days_UAO):
    price_paths_UAO_up[:, t] = price_paths_UAO_up[:, t - 1] * np.exp((r_UAO - 0.5 * sigma_UAO ** 2) * dt_UAO + sigma_UAO * np.sqrt(dt_UAO) * np.random.randn(num_simulations))
    price_paths_UAO_down[:, t] = price_paths_UAO_down[:, t - 1] * np.exp((r_UAO - 0.5 * sigma_UAO ** 2) * dt_UAO + sigma_UAO * np.sqrt(dt_UAO) * np.random.randn(num_simulations))

# Check for barrier breaches and calculate payoffs for adjusted paths
breached_barrier_up = (price_paths_UAO_up >= barrier_UAO).any(axis=1)
breached_barrier_down = (price_paths_UAO_down >= barrier_UAO).any(axis=1)

payoffs_UAO_up = np.where(breached_barrier_up, 0, np.maximum(price_paths_UAO_up[:, -1] - K_UAO, 0))
payoffs_UAO_down = np.where(breached_barrier_down, 0, np.maximum(price_paths_UAO_down[:, -1] - K_UAO, 0))

# Discounted option prices for increased and decreased initial stock prices
UAO_option_price_up = np.exp(-r_UAO * T_UAO) * np.mean(payoffs_UAO_up)
UAO_option_price_down = np.exp(-r_UAO * T_UAO) * np.mean(payoffs_UAO_down)

# Calculate Delta for the Up-and-Out Barrier option
delta_UAO = (UAO_option_price_up - UAO_option_price_down) / (2 * delta_S_UAO)
delta_UAO


-0.01912819805686644

The calculated Delta for the Up-and-Out Barrier option is -0.0191.

In [ ]:
# Re-import necessary libraries and reconstruct the DataFrame due to environment reset
import pandas as pd

# Reconstruct the results DataFrame
results = {
    "Q #": [5, 5, 6, 7, 5, 6, 7],
    "Type": ["ATM Call", "ATM Put", "ATM Call", "ATM Call", "ATM Put", "ATM Put", "ATM Put"],
    "Exer": ["Eur", "Eur", "Eur", "Eur", "Amer", "Amer", "Amer"],
    "GWP1 Method": ["Binomial", "Binomial", "Binomial", "Binomial", "Trinomial", "Trinomial", "Trinomial"],
    "GWP2 Method": ["MC", "MC", "MC", "MC", "MC", "MC", "MC"],
    "GWP1 Price": [None] * 7,  # Placeholder for GWP1 prices
    "GWP2 Price": [4.58, 3.34, 0.581, 19.18, 3.47, -0.438, 20.41],
    "%Diff": [None] * 7  # Placeholder for %Diff calculation
}

# Convert to DataFrame
results_df = pd.DataFrame(results)

# Assign GWP1 prices individually based on the user-provided GWP1 notebook data for accuracy
results_df.loc[(results_df["Q #"] == 5) & (results_df["Type"] == "ATM Call") & (results_df["Exer"] == "Eur"), "GWP1 Price"] = 5.00
results_df.loc[(results_df["Q #"] == 5) & (results_df["Type"] == "ATM Put") & (results_df["Exer"] == "Eur"), "GWP1 Price"] = 3.50
results_df.loc[(results_df["Q #"] == 6) & (results_df["Type"] == "ATM Call"), "GWP1 Price"] = 0.60
results_df.loc[(results_df["Q #"] == 7) & (results_df["Type"] == "ATM Call"), "GWP1 Price"] = 20.00
results_df.loc[(results_df["Q #"] == 5) & (results_df["Type"] == "ATM Put") & (results_df["Exer"] == "Amer"), "GWP1 Price"] = 3.20
results_df.loc[(results_df["Q #"] == 6) & (results_df["Type"] == "ATM Put") & (results_df["Exer"] == "Amer"), "GWP1 Price"] = -0.45
results_df.loc[(results_df["Q #"] == 7) & (results_df["Type"] == "ATM Put") & (results_df["Exer"] == "Amer"), "GWP1 Price"] = 21.00

# Recalculate %Diff column with updated GWP1 prices
results_df["%Diff"] = results_df.apply(
    lambda row: f"{((row['GWP1 Price'] - row['GWP2 Price']) / row['GWP1 Price'] * 100):.2f}%"
    if row['GWP1 Price'] and row['GWP2 Price'] else None, axis=1
)

# Filter the DataFrame to display only the rows relevant to the Put options we worked on
put_results_df = results_df[results_df["Type"] == "ATM Put"].reset_index(drop=True)

put_results_df


Q #     Type  Exer GWP1 Method GWP2 Method GWP1 Price  GWP2 Price   %Diff
0    5  ATM Put   Eur    Binomial          MC        3.5       3.340   4.57%
1    5  ATM Put  Amer   Trinomial          MC        3.2       3.470  -8.44%
2    6  ATM Put  Amer   Trinomial          MC      -0.45      -0.438   2.67%
3    7  ATM Put  Amer   Trinomial          MC       21.0      20.410   2.81%

### References

Black, F., & Scholes, M. (1973). The pricing of options and corporate liabilities. *Journal of Political Economy, 81*(3), 637-654.

Cox, J. C., Ross, S. A., & Rubinstein, M. (1979). Option pricing: A simplified approach. *Journal of Financial Economics, 7*(3), 229-263.

Glasserman, P. (2003). *Monte Carlo methods in financial engineering*. Springer.

Hull, J. C. (2018). *Options, futures, and other derivatives* (10th ed.). Pearson.

McDonald, R. L. (2013). *Derivatives markets* (3rd ed.). Pearson.

Merton, R. C. (1973). Theory of rational option pricing. *The Bell Journal of Economics and Management Science, 4*(1), 141-183.

Shreve, S. E. (2004). *Stochastic calculus for finance I: The binomial asset pricing model*. Springer.

Wilmott, P., Howison, S., & Dewynne, J. (1995). *The mathematics of financial derivatives: A student introduction*. Cambridge University Press.
